File: src/logic/pieces/bishop.py
from logic.move import Move
from .piece import Piece

class Bishop(Piece):
    def legal_moves(self, board: "Board", / , looking_for_check = False) -> list[Move]:
        ret = []

        # looking north east
        ret.extend(self._look_direction(board, 1, 1))

        # looking south east
        ret.extend(self._look_direction(board, 1, -1))

        # looking south west
        ret.extend(self._look_direction(board, -1, -1))

        # looking north west
        ret.extend(self._look_direction(board, -1, 1))

        if not looking_for_check and board.is_check_for(self.colour):
            return self.keep_only_blocking(ret, board)

        return ret

File: cpp/src/pieces/bishop.cpp
#include "../board.hpp"
#include "../coords.hpp"
#include "../move.hpp"
#include "piece.hpp"

#include <vector>

std::vector<Move> bishop_moves(const Board& b, const Coords xy) {
    std::vector<Move> ret;
    auto ne = look_direction(b, xy, 1, 1);
    ret.insert(ret.end(), ne.begin(), ne.end());

    auto se = look_direction(b, xy, 1, -1);
    ret.insert(ret.end(), se.begin(), se.end());

    auto sw = look_direction(b, xy, -1, -1);
    ret.insert(ret.end(), sw.begin(), sw.end());

    auto nw = look_direction(b, xy, -1, 1);
    ret.insert(ret.end(), nw.begin(), nw.end());

    return ret;
}
